### DISCUSSION TITLE
Test Environment for API Usage

### DISCUSSION THREAD

**Discussion by Danielku15** [community/community#24392] [Enterprise]
<p>Hi everyone.</p>
<p><strong>tl;dr:</strong> How is it meant to test applications that use the GitHub API? Is there some test environment which is reset regularly, can I start an own environment, or shall I “spam” <a href="http://GitHub.com" rel="nofollow noopener">GitHub.com</a> with some test data for my purposes?</p>
<p><strong>long version:</strong><br>
We lately bought GitHub Enterprise for the usage in our company. Our department got an own organization within it for our usage across the teams we have. Currently we’re administering all teams, repositories, memberships etc. manually via the GitHub Web Interface. This causes some troubles due to inconsistencies in naming conventions, and team memberships and people just not being familiar with GitHub and the policies we have.</p>
<p>So what I would like to do, is to move to a Terraform based administration (Infrastructure As A Code). Terraform has a GitHub provider which would allow us to manage the organization in a proper code workflow via PRs, Checks and Code Reviews. Instead of having many admins for all teams, they would just file a PR on the configuration repo where it is then approved or rejected. Some training will be needed here of course but having some examples/templates should solve this quickly.</p>
<p>The unfortunate thing is that I do not have a test environment where I could start learning about how this could work. I could register a test organization on <a href="http://GitHub.com" rel="nofollow noopener">GitHub.com</a> or ask the admins of our GHE to give me one temporarily. But then I’m still without users to test.</p>
<p>Shall I just register a bunch of test users in <a href="http://GitHub.com" rel="nofollow noopener">GitHub.com</a> for my purpose (will need likely around 10)? I could also ask some people but it just does not work out that smooth for any testing. Is there maybe a dedicated <a href="http://GitHub.com" rel="nofollow noopener">GitHub.com</a> test environment that could be used or can I spin-up some local API environment for this? I would be even willing to just fire up a local trial GitHub Enterprise instance in a VM if this is possible.</p>
<p>I could not really find any recommendation or references on this topic in the docs.<br>
<a href="https://docs.github.com/en/developers/apps/setting-up-your-development-environment-to-create-a-github-app" rel="nofollow noopener">This page</a> shows some info about registering an app in your own account, but that’s more a WebHook based App which reacts on stuff happening in GH while I want to administer it.<br>
For a GHE trial I need to <a href="https://docs.github.com/en/github/getting-started-with-github/setting-up-a-trial-of-github-enterprise-server" rel="nofollow noopener">get in contact with Sales</a> which also gives me just 45 days to get started (not much if you just do this a side task).</p>
<p>Looking forward to get some input here.</p>
<p>Cheers<br>
Daniel</p>

**Comment by francisfuzz** [comment, marked as answer]
<p>:wave: <a class="mention" href="/Danielku15">@Danielku15</a>: Hello there! Thanks for sharing these questions and that additional context. While I can’t speak to the Terraform portion, I’d like to share some input towards the GitHub API and GitHub Enterprise items.</p>
<aside class="quote no-group" data-username="Danielku15" data-post="1" data-topic="124894">
<div class="title">
<div class="quote-controls"></div>
 Danielku15:</div>
<blockquote>
<p><strong>tl;dr:</strong> How is it meant to test applications that use the GitHub API? Is there some test environment which is reset regularly, can I start an own environment, or shall I “spam” <a href="http://GitHub.com">GitHub.com</a> with some test data for my purposes?</p>
</blockquote>
</aside>
<p>We are absolutely thrilled that you’d like to use our GitHub API for your workflow! :tada:</p>
<p>On <a href="http://GitHub.com">GitHub.com</a>, there is not a specific “area” or feature for spinning up a test environment to make test API calls. However, it is possible to have a “staging” area in GitHub Enterprise (more on that below).</p>
<p>If you’re interested in making some test calls to the <a href="http://GitHub.com">GitHub.com</a> API, we suggest <a href="https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token">creating a personal access token</a> and using it to <a href="https://docs.github.com/en/rest/overview/resources-in-the-rest-api#oauth2-token-sent-in-a-header">authenticate</a>. We have a guide for <a href="https://docs.github.com/en/rest/guides/getting-started-with-the-rest-api">getting started with the REST API</a> as well as an extensive <a href="https://docs.github.com/en/rest/reference">reference</a> for all of the resources available for this interface.</p>
<p>As a new integrator, we have a set of <a href="https://docs.github.com/en/rest/guides/best-practices-for-integrators">best practices for integrators</a> that’s intended to help make the most of your experience on our platform. You’re welcome to create a <a href="https://docs.github.com/en/github/site-policy/github-terms-of-service#3-account-requirements">machine account</a> that’s used exclusively for performing automated tasks (including interfacing with the GitHub API). We discourage using the API for “spamming” purposes, and you’re welcome to read more about our <a href="https://docs.github.com/en/github/site-policy/github-terms-of-service#h-api-terms">API terms</a> regarding request usage.</p>
<aside class="quote no-group" data-username="Danielku15" data-post="1" data-topic="124894">
<div class="title">
<div class="quote-controls"></div>
 Danielku15:</div>
<blockquote>
<p>The unfortunate thing is that I do not have a test environment where I could start learning about how this could work. I could register a test organization on <a href="http://GitHub.com">GitHub.com</a> or ask the admins of our GHE to give me one temporarily. But then I’m still without users to test.</p>
<p>Shall I just register a bunch of test users in <a href="http://GitHub.com">GitHub.com</a> for my purpose (will need likely around 10)? I could also ask some people but it just does not work out that smooth for any testing. Is there maybe a dedicated <a href="http://GitHub.com">GitHub.com</a> test environment that could be used or can I spin-up some local API environment for this? I would be even willing to just fire up a local trial GitHub Enterprise instance in a VM if this is possible.</p>
<p>I could not really find any recommendation or references on this topic in the docs.<br>
<a href="https://docs.github.com/en/developers/apps/setting-up-your-development-environment-to-create-a-github-app">This page</a> shows some info about registering an app in your own account, but that’s more a WebHook based App which reacts on stuff happening in GH while I want to administer it.<br>
For a GHE trial I need to <a href="https://docs.github.com/en/github/getting-started-with-github/setting-up-a-trial-of-github-enterprise-server">get in contact with Sales</a> which also gives me just 45 days to get started (not much if you just do this a side task).</p>
</blockquote>
</aside>
<p>If your GHE administrators already have access to an existing GitHub Enterprise instance, it’s possible to <a href="https://docs.github.com/en/enterprise/2.20/admin/installation/setting-up-a-staging-instance">setup a staging instance</a> for testing, including make requests to the API. As a next step, we suggest sharing that with them as a part of your testing; if they have any questions, they can <a href="https://enterprise.githubsupport.com/hc/en-us">open a new support ticket using the GitHub Enterprise Support portal</a>.</p>
<p>Does this help with what you’re looking for as far as next steps go? :thought_balloon:</p>

**Comment by Danielku15** [comment]
<p><a class="mention" href="/francisfuzz">@francisfuzz</a> Thanks for your reply. It is unfortunate to hear that there seems no straight forward way to do these kind of tests against the GitHub API. Repository and Team management can be done with a single account. But as soon as it comes to inviting people to orgs and assigning them to teams with permissions I will need some other accounts.</p>
<p>From an actual authentication and “getting started with the API” everything should be quite clear. The API docs there are very good and the main thing to learn will be to use Terraform together with GH. :nerd_face:</p>
<p>It is unlikely that within a company of our size there is a chance that our GHE admins (our IT) will setup a staging instance for me to test things. Such an environment will result in related costs (setting up, server operations etc.) which usually result in a decline of the request. I would do it on my own, but due to the lack of a license key I cannot really boot it up as our IT will not give me our company key. :sob:</p>
<p>I might rather ask some friends with accounts on <a href="http://GitHub.com" rel="nofollow noopener">GitHub.com</a> and setup a public org for testing the permission management. Of course it would be more convenient to have some personal test accounts which I just delete afterwards once things are working. :thinking:</p>
